### PR TITLE
ReadMeta function to get the CollectionType

### LINKED ISSUE #174: Add Recognition for new DCDataTypes
We need to add enums and the general recognition in Splash for the new data types:

- DCDT_BOOL: #153 #198
- DCDT_CHAR, DCDT_STRING, DCDT_VLSTRING: #167
- ARRAY and COMPOUNDS?

- [x] At least [the following files](https://github.com/ax3l/libSplash/commit/a34235ccf75466ec3fa4959c81d2edbc8e0e4bc3) need an update.
- [x] Also, tests should be added that actually use `DCDataSet::getDCDataType()` and `DomainData::getFormatString()` (probably, both are only used in `splash2txt` -> splashtools)

### PR DESCRIPTION
**Pull request by slizzered**
Closes #174 and #189 based on the draft commit mentioned in https://github.com/ComputationalRadiationPhysics/libSplash/issues/189, a type deduction system was implemented.

Implements most of #196 for testing.

~~There is still some (minor) TODO left, see line 568-570 in `src/ParallelDataCollector.cpp`~~

### CODE REVIEW

File: src/ParallelDataCollector.cpp [hunk 1]
```diff
@@ -559,6 +559,22 @@ namespace splash
...
     }
 
+     CollectionType* ParallelDataCollector::readMeta(int32_t id,
+             const char* name,
+             const Dimensions dstBuffer,
+             const Dimensions dstOffset,
+             Dimensions& sizeRead) throw (DCException)
+     {
+         // TODO: Check what this line does (FST_MERGING is not defined!)
+//       if (fileStatus != FST_READING && fileStatus != FST_WRITING && fileStatus != FST_MERGING)
```
Comment (ax3l) [inline src/ParallelDataCollector.cpp]: should basically check if the PDC object is valid and a file is opened.

You can copy that from `ParallelDataCollector::read`

File: src/ParallelDataCollector.cpp [hunk 2]
```diff
@@ -559,6 +559,20 @@ namespace splash
...
     }
 
+     CollectionType* ParallelDataCollector::readMeta(int32_t id,
+             const char* name,
+             const Dimensions dstBuffer,
+             const Dimensions dstOffset,
+             Dimensions& sizeRead) throw (DCException)
+     {
+         if (fileStatus != FST_READING && fileStatus != FST_WRITING)
+             throw DCException(getExceptionString("read", "this access is not permitted"));
```
Comment (f-schmitt-zih) [inline src/ParallelDataCollector.cpp]: fix the function name in the exception

File: src/ParallelDataCollector.cpp [hunk 3]
```diff
@@ -559,6 +559,20 @@ namespace splash
...
+
+         uint32_t ndims = 0;
+         return readDataSetMeta(handles.get(0), id, name, dstBuffer, dstOffset,
```
Comment (f-schmitt-zih) [inline src/ParallelDataCollector.cpp]: why always accessing handle 0?

File: src/ParallelDataCollector.cpp [hunk 4]
```diff
@@ -956,6 +970,45 @@ namespace splash
...
+            const char* name,
+            const Dimensions dstBuffer,
+            const Dimensions dstOffset,
+            const Dimensions srcOffset,
+            Dimensions &sizeRead,
+            uint32_t& srcDims)
+    throw (DCException)
+    {
+        log_msg(2, "readDataSetMeta");
+
```
Comment (f-schmitt-zih) [inline src/ParallelDataCollector.cpp]: as this is a public function, you should at least check h5File and name parameters

File: src/ParallelDataCollector.cpp [hunk 5]
```diff
@@ -956,6 +970,45 @@ namespace splash
...
+        std::string group_path, dset_name;
+        DCDataSet::getFullDataPath(name, SDC_GROUP_DATA, id, group_path, dset_name);
+
+        DCGroup group;
+        group.open(h5File, group_path);
+
+        DCDataSet dataset(dset_name.c_str());
+        dataset.open(group.getHandle());
+
+        //size_t datatype_size = 0;
```
Comment (f-schmitt-zih) [inline src/ParallelDataCollector.cpp]: please remove dead code

File: src/ParallelDataCollector.cpp [hunk 6]
```diff
@@ -956,6 +970,45 @@ namespace splash
...
+        //DCDataType dc_datatype = DCDT_UNKNOWN;
+        //datatype_size = dataset.getDataTypeSize();
+        //dc_datatype = dataset.getDCDataType();
+
+        size_t entrySize;
+        getEntriesForID(id, NULL, &entrySize);
+        std::vector<DataCollector::DCEntry> entries(entrySize);
+
+        getEntriesForID(id, &(*entries.begin()), NULL);
```
Comment (f-schmitt-zih) [inline src/ParallelDataCollector.cpp]: what is the &(*...) for?

File: src/ParallelDataCollector.cpp [hunk 7]
```diff
@@ -956,6 +970,45 @@ namespace splash
...
+
+        Dimensions src_size(dataset.getSize() - srcOffset);
+        dataset.read(dstBuffer, dstOffset, src_size, srcOffset, sizeRead, srcDims, NULL);
+        dataset.close();
+
+        return entries[0].col_type;
```
Comment (f-schmitt-zih) [inline src/ParallelDataCollector.cpp]: could entries be empty and not have a 0th element?

File: src/SerialDataCollector.cpp [hunk 8]
```diff
@@ -425,6 +426,21 @@ namespace splash
...
 
+    CollectionType* SerialDataCollector::readMeta(int32_t id,
+            const char* name,
+            const Dimensions dstBuffer,
+            const Dimensions dstOffset,
+            Dimensions &sizeRead)
+    throw (DCException)
+    {
+        if (fileStatus != FST_READING && fileStatus != FST_WRITING && fileStatus != FST_MERGING)
+            throw DCException(getExceptionString("read", "this access is not permitted"));
```
Comment (f-schmitt-zih) [inline src/SerialDataCollector.cpp]: fix function name

File: src/SerialDataCollector.cpp [hunk 9]
```diff
@@ -425,6 +426,21 @@ namespace splash
...
+
+        uint32_t ndims = 0;
+        return readDataSetMeta(handles.get(0), id, name, dstBuffer, dstOffset,
```
Comment (f-schmitt-zih) [inline src/SerialDataCollector.cpp]: same here: handle 0?

File: src/SerialDataCollector.cpp [hunk 10]
```diff
@@ -957,6 +973,45 @@ namespace splash
...
+            const char* name,
+            const Dimensions dstBuffer,
+            const Dimensions dstOffset,
+            const Dimensions srcOffset,
+            Dimensions &sizeRead,
+            uint32_t& srcDims)
+    throw (DCException)
+    {
+        log_msg(2, "readDataSetMeta");
+
```
Comment (f-schmitt-zih) [inline src/SerialDataCollector.cpp]: check parameters

File: src/SerialDataCollector.cpp [hunk 11]
```diff
@@ -957,6 +973,45 @@ namespace splash
...
+        std::string group_path, dset_name;
+        DCDataSet::getFullDataPath(name, SDC_GROUP_DATA, id, group_path, dset_name);
+
+        DCGroup group;
+        group.open(h5File, group_path);
+
+        DCDataSet dataset(dset_name.c_str());
+        dataset.open(group.getHandle());
+
+        //size_t datatype_size = 0;
```
Comment (f-schmitt-zih) [inline src/SerialDataCollector.cpp]: remove

File: src/include/splash/DataCollector.hpp [hunk 12]
```diff
@@ -113,6 +114,8 @@ namespace splash
              * Fully-qualified name of this dataset.
              */
             std::string name;
+            DCDataType dc_type;
```
Comment (f-schmitt-zih) [inline src/include/splash/DataCollector.hpp]: comment these members and stick to the existing naming convention (i.e. no "_")

File: src/include/splash/ParallelDataCollector.hpp [hunk 13]
```diff
@@ -312,6 +312,12 @@ namespace splash
                 Dimensions &sizeRead,
                 void* buf) throw (DCException);
 
+        CollectionType* readMeta(int32_t id,
```
Comment (f-schmitt-zih) [inline src/include/splash/ParallelDataCollector.hpp]: this functions needs documentation

File: src/include/splash/ParallelDataCollector.hpp [hunk 14]
```diff
@@ -408,6 +414,17 @@ namespace splash
                 Dimensions /*count*/,
                 Dimensions /*offset*/,
                 Dimensions /*stride*/) throw (DCException);
+
+
+        CollectionType* readDataSetMeta(H5Handle h5File,
```
Comment (f-schmitt-zih) [inline src/include/splash/ParallelDataCollector.hpp]: this function needs documentation

File: src/include/splash/SerialDataCollector.hpp [hunk 15]
```diff
@@ -161,6 +161,17 @@ namespace splash
                 void* dst)
         throw (DCException);
 
+        void readDataSetMeta(H5Handle h5File,
```
Comment (f-schmitt-zih) [inline src/include/splash/SerialDataCollector.hpp]: this function needs documentation

File: src/include/splash/SerialDataCollector.hpp [hunk 16]
```diff
@@ -175,6 +186,15 @@ namespace splash
                 uint32_t& srcDims,
                 void* dst) throw (DCException);
 
+        CollectionType* readDataSetMeta(H5Handle h5File,
```
Comment (f-schmitt-zih) [inline src/include/splash/SerialDataCollector.hpp]: ...

File: src/include/splash/SerialDataCollector.hpp [hunk 17]
```diff
@@ -321,6 +341,12 @@ namespace splash
                 const Dimensions dstOffset,
                 Dimensions &sizeRead,
                 void* data) throw (DCException);
+
+        CollectionType* readMeta(int32_t id,
```
Comment (f-schmitt-zih) [inline src/include/splash/SerialDataCollector.hpp]: add documentation

File: src/include/splash/basetypes/ColTypeBool.hpp [hunk 18]
```diff
@@ -25,6 +26,10 @@
 #define	COLTYPEBOOL_H
 
 #include "splash/CollectionType.hpp"
+#include "splash/basetypes/ColTypeUnknown.hpp"
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/ColTypeBool.hpp]: what is the ColTypeUnknown include for? (note that there are multiple occurences in other type headers)

File: src/include/splash/basetypes/ColTypeBool.hpp [hunk 19]
```diff
@@ -55,6 +60,32 @@ class ColTypeBool : public CollectionType
     {
         return sizeof (bool);
     }
+
+    static CollectionType* genType(hid_t datatype_id)
+    {
+        bool found = false;
+        H5T_class_t h5_class = H5Tget_class(datatype_id);
+        if(h5_class == H5T_ENUM){
+            if(H5Tget_nmembers(datatype_id) == 2){
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/ColTypeBool.hpp]: please stick to the existing "parantheses on next line" style everywhere

File: src/include/splash/basetypes/ColTypeDim.hpp [hunk 20]
```diff
@@ -51,6 +56,35 @@ namespace splash
         {
             return Dimensions::getSize();
         }
+
+        std::string toString() const
+        {
+            return "Dimension";
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/ColTypeDim.hpp]: ColTypeDimArray is "DimArray" but ColTypeDim is "Dimension", you should decide on one style

File: src/include/splash/basetypes/ColTypeDim.hpp [hunk 21]
```diff
@@ -51,6 +56,35 @@ namespace splash
...
+        {
+            bool found = false;
+            H5T_class_t h5_class = H5Tget_class(datatype_id);
+            if(h5_class == H5T_COMPOUND){
+                if(H5Tget_nmembers(datatype_id) == 3){
+                    if(H5Tget_size(datatype_id) == Dimensions::getSize()){
+                        char* m0 = H5Tget_member_name(datatype_id, 0);
+                        char* m1 = H5Tget_member_name(datatype_id, 1);
+                        char* m2 = H5Tget_member_name(datatype_id, 2);
+                        if( (strcmp("x", m0) || strcmp("y", m1) || strcmp("z", m2)) == 0)
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/ColTypeDim.hpp]: there is no reason to outsmart other programmers. use "==0" for all individual checks as this is what you mean

File: src/include/splash/basetypes/ColTypeDimArray.hpp [hunk 22]
```diff
@@ -47,6 +50,35 @@ namespace splash
...
         }
+
+        static CollectionType* genType(hid_t datatype_id)
+        {
+            bool found = false;
+            H5T_class_t h5_class = H5Tget_class(datatype_id);
+            if(h5_class == H5T_ARRAY)
+            {
+                hid_t base = H5Tget_super(datatype_id);
+                if(H5Tequal(base, H5T_NATIVE_HSIZE) == 1){
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/ColTypeDimArray.hpp]: style

File: src/include/splash/basetypes/ColTypeString.hpp [hunk 23]
```diff
@@ -70,6 +73,31 @@ namespace splash
             else
                return sizeof(char) * (myElements - 1); /* just as strlen() */
         }
+
+        static CollectionType* genType(hid_t datatype_id)
+        {
+            H5T_class_t h5_class = H5Tget_class(datatype_id);
+
+            if(h5_class == H5T_STRING){
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/ColTypeString.hpp]: style everywhere

File: src/include/splash/basetypes/CollectionTypeGenerator.hpp [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+#define COLTYPEGENERATOR_H
+
+
+#include "splash/basetypes/basetypes.hpp"
+#include "splash/CollectionType.hpp"
+
+namespace splash
+{
+
+#define COLLECTION_TEST(_name)                                                 \
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/CollectionTypeGenerator.hpp]: please add a short comment what this macro is doing

File: src/include/splash/basetypes/CollectionTypeGenerator.hpp [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+{                                                                              \
+    CollectionType* t = ColType##_name::genType(datatype_id);                  \
+    if(t != NULL && typeid(*t) == typeid(ColType##_name)) return t;            \
+}                                                                              \
+
+
+class CollectionTypeGenerator{
+public:
+
+    static CollectionType* genCollectionType(hid_t datatype_id)
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/CollectionTypeGenerator.hpp]: Consider making this a non-member function, as it does not seem to be related to CollectionTypeGenerator class.
Add a comment on the return, i.e. does the caller have to free/delete it?

File: src/include/splash/SerialDataCollector.hpp (hunk 15 above)
Comment (ax3l) [inline src/include/splash/SerialDataCollector.hpp]: also: wrong return type

File: src/SerialDataCollector.cpp (hunk 8 above)
Comment (ax3l) [inline src/SerialDataCollector.cpp]: "read" -> "readMeta"

File: src/SerialDataCollector.cpp (hunk 10 above)
Comment (ax3l) [inline src/SerialDataCollector.cpp]: that's a private member such as `readDataSet` and state of the `SerialDataCollector` (open, readable, ...) has already been checked at this point

File: src/include/splash/SerialDataCollector.hpp (hunk 16 above)
Comment (ax3l) [inline src/include/splash/SerialDataCollector.hpp]:
```
/**
 * Internal meta data reading method.
 */
```

File: src/ParallelDataCollector.cpp (hunk 2 above)
Comment (ax3l) [inline src/ParallelDataCollector.cpp]: read -> readMeta

File: src/ParallelDataCollector.cpp (hunk 3 above)
Comment (ax3l) [inline src/ParallelDataCollector.cpp]: huh, should be `handles.get(id)` I guess

File: src/ParallelDataCollector.cpp (hunk 4 above)
Comment (ax3l) [inline src/ParallelDataCollector.cpp]: this should be an `private` member and is only called by members that already checked the file's state.

File: src/ParallelDataCollector.cpp (hunk 6 above)
Comment (ax3l) [inline src/ParallelDataCollector.cpp]: C++98 version of [`std::vector::data()`](http://www.cplusplus.com/reference/vector/vector/data/): *iterator* of first element -> dereference -> get pointer of actual data element
(iterator != pointer of first element in storage; C++ `std::vector`s guarantees a contiguous C-like array internally)

File: src/ParallelDataCollector.cpp [hunk 26]
```diff
@@ -956,6 +970,45 @@ namespace splash
...
+        getEntriesForID(id, NULL, &entrySize);
+        std::vector<DataCollector::DCEntry> entries(entrySize);
+
+        getEntriesForID(id, &(*entries.begin()), NULL);
+
+        Dimensions src_size(dataset.getSize() - srcOffset);
+        dataset.read(dstBuffer, dstOffset, src_size, srcOffset, sizeRead, srcDims, NULL);
+        dataset.close();
+
+        return entries[0].colType;
```
Comment (ax3l) [inline src/ParallelDataCollector.cpp]: @f-schmitt-zih that's a bit hacky but if I remember correctly, should work since they all have the same type.

File: src/SerialDataCollector.cpp (hunk 9 above)
Comment (ax3l) [inline src/SerialDataCollector.cpp]: `handles.get(id)`

File: src/include/splash/ParallelDataCollector.hpp (hunk 13 above)
Comment (ax3l) [inline src/include/splash/ParallelDataCollector.hpp]: not sure if relevant, but `read` is not documented either at this place but just in the `DataCollector` interface (?)

Comment (ax3l) [inline src/include/splash/ParallelDataCollector.hpp]: -> needs a virtual interface and documentation in `DataCollector`

File: src/include/splash/basetypes/generateCollectionType.hpp [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+ * Copyright 2015 Carlchristian Eckert
+ *
+ * This file is part of libSplash.
+ *
+ * libSplash is free software: you can redistribute it and/or modify
+ * it under the terms of of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * libSplash is distributed in the hope that it will be useful,
```
Comment (ax3l) [inline src/include/splash/basetypes/generateCollectionType.hpp]: newline

File: src/include/splash/ParallelDataCollector.hpp (hunk 13 above)
Comment (ax3l) [inline src/include/splash/ParallelDataCollector.hpp]: done

File: src/include/splash/SerialDataCollector.hpp (hunk 17 above)
Comment (ax3l) [inline src/include/splash/SerialDataCollector.hpp]: as the members above, directly documented in the `DataCollector` interface

File: src/ParallelDataCollector.cpp [hunk 28]
```diff
@@ -956,6 +970,40 @@ namespace splash
...
```
Comment (ax3l) [inline src/ParallelDataCollector.cpp]: @f-schmitt-zih that's a bit hacky but if I remember correctly, should work since they all have the same type.

File: src/SerialDataCollector.cpp [hunk 29]
```diff
@@ -957,6 +973,40 @@ namespace splash
...
+        getEntriesForID(id, NULL, &entrySize);
+        std::vector<DataCollector::DCEntry> entries(entrySize);
+
+        getEntriesForID(id, &(*entries.begin()), NULL);
+
+        Dimensions src_size(dataset.getSize() - srcOffset);
+        dataset.read(dstBuffer, dstOffset, src_size, srcOffset, sizeRead, srcDims, NULL);
+        dataset.close();
+
+        return entries[0].colType;
```
Comment (ax3l) [inline src/SerialDataCollector.cpp]: @f-schmitt-zih that's a bit hacky but if I remember correctly, should work since they all have the same type.

File: src/ParallelDataCollector.cpp (hunk 4 above)
Comment (f-schmitt-zih) [inline src/ParallelDataCollector.cpp]: I see,
```
/*******************************************************************************
* PUBLIC FUNCTIONS
*******************************************************************************/
```
means nothing anymore these days...
(just joking, good point)

Comment (ax3l) [inline src/ParallelDataCollector.cpp]: there is such a gentle block in L160 but L806 did already announce the `protected party zone - here will be dragons`.

Comment (ax3l) [inline src/ParallelDataCollector.cpp]: probably it would be easier to read in our venerable age if we would add more ascii art to the blocks. kids these day...!

Comment (f-schmitt-zih) [inline src/ParallelDataCollector.cpp]: You should go ahead and document all function behavior using ascii art flow
charts... like a pro

Comment (ax3l) [inline src/ParallelDataCollector.cpp]:
```
 +--------+   +     +   +-+
 |        |   |   --+   |-|
 |        |   | |       |-+
 |        |   +-+--+    ++
 |        |   |    |
 +--------+   +    ++   ++

 XXXXX      XXXX       XXX
XX   XX  XXXX  XXX    XX XXX
      XXXX        XXXXX    XXXXX
```

But [asciiflow.com](http://asciiflow.com/) took even that joy away from us making it a less-then-a-few-months task :rabbit:

File: src/ParallelDataCollector.cpp (hunk 28 above)
Comment (ax3l) [inline src/ParallelDataCollector.cpp]: @slizzered I just worked with this function and there is a bug in this return statement.

`getEntriesForID` gets *all* data sets inside this `id` (time step) but you should *filter* for the `id` `i` that is `name == entries[i].name`

File: src/SerialDataCollector.cpp (hunk 29 above)
Comment (ax3l) [inline src/SerialDataCollector.cpp]: same bug as in Parallel reader
